cabot/cabotapp/tests/test_api.py

```diff
@@ -492,6 +501,99 @@ def test_check_should_not_run_when_activity_counter_missing(self):
+        self.assertTrue(self.http_check.should_run())
+
+    def test_check_should_run_should_set_last_enabled_if_null(self):
+        # This tests the behavior of should_run(), and under what circumstances
+        # it should set the last_enabled field.
+        #
+        # NB: we use self._get_activity_counter() instead of just saving the counter
+        # as a local variable, because we want to test the latest values in the DB.
+
+        # Create the activity_counter
+        self._set_activity_counter(True, 0)
+
+        # Make sure last_enabled is None
+        self.assertTrue(self.http_check.use_activity_counter)
+        self.assertIsNone(self._get_activity_counter().last_enabled)
+
+        # Even though use_activity_counter is True, should_run() will not set last_enabled
+        # because the count is 0 and last_enabled is None (a special case).
+        self.assertFalse(self.http_check.should_run())
+        self.assertIsNone(self._get_activity_counter().last_enabled)
+
+        # Once count is positive and we need last_enabled, it'll be set
+        counter = self._get_activity_counter()
+        counter.count = 4
+        counter.save()
+        self.assertTrue(self.http_check.should_run())
+        self.assertIsNotNone(self._get_activity_counter().last_enabled)
+
+    @patch('cabot.cabotapp.models.timezone.now')
+    def test_check_should_run_with_delay(self, mock_now):
+        # Use the following times:
+        #
+        # - last_enabled  = T+0
+        # - last_disabled = T+60
+        # - run_delay     = 30 mins
+        #
+        # The check should thus run between T+30 and T+90
+        self._set_activity_counter(True, 1)
+        self.http_check.run_delay = 30
+        self.http_check.save()
+        counter = self._get_activity_counter()
+
+        # Assume it is 15 mins after check was enabled. Check should not run right now.
+        mock_now.return_value = counter.last_enabled + timedelta(minutes=15)
+        self.assertFalse(self.http_check.should_run())
+
+        # Check should run if now >= run_delay (30 minutes)
+        mock_now.return_value = counter.last_enabled + timedelta(minutes=30)
+        self.assertTrue(self.http_check.should_run())
+
+        # Now assume the check ran recently. It shouldn't run again until the frequency is up.
+        self.http_check.last_run = counter.last_enabled + timedelta(minutes=31)
+        self.http_check.save()
+
+        mock_now.return_value = counter.last_enabled + timedelta(minutes=32)
+        self.assertFalse(self.http_check.should_run())
+
+        mock_now.return_value = counter.last_enabled + timedelta(minutes=37)
+        self.assertTrue(self.http_check.should_run())
+
+        # Finally, reset the counter and set the last_disabled time to last_enabled + 60.
+        # The check should not run if after the disabled time (last_enabled + 90).
+        counter.last_disabled = counter.last_enabled + timedelta(minutes=60)
```

Comment:
Consider using the `counter.decrement()` method to better simulate what the decrement API would do? (Or even use `self.client.get(...)`)